File: apps/mobile/android/shared/src/commonMain/kotlin/cc/edecan/shared/AutomationsModels.kt
package cc.edecan.shared

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Modelos de `/v1/automations` (`ROADMAP_V2.md` §7.4, §7.6, §7.10;
 * `apps/api/edecan_api/routers/automations.py::_public_automation`) —
 * pestaña "Automatizaciones" (WP-V5-07). Mismo criterio de defaults
 * tolerantes que [MissionsModels.kt]/`Models.kt`.
 */

/** `edecan_schemas.plans.FLAG_AUTOMATIONS_RULES` — gate de plan de TODO
 * `/v1/automations` (`automations.py::_require_automations_flag`). */
const val FLAG_AUTOMATIONS_RULES = "automations.rules"

/**
 * `Automation.trigger` — las dos variantes de `TriggerDef`
 * (`edecan_schemas.automations`, unión discriminada por `kind` del lado del
 * backend) aplanadas en un solo data class tolerante en vez de un sellado:
 * `rrule` solo viene poblado si `kind == "schedule"`; `hasSecret`/`hookUrl`
 * solo si `kind == "webhook"` — **nunca** `hook_secret` en claro aquí (ver
 * el docstring de `_public_automation`: se redacta a `has_secret`/`hook_url`
 * en todo `GET`). Este cliente móvil (WP-V5-07) solo ofrece crear
 * automatizaciones `kind == "schedule"` (ver `AutomatizacionesScreen`,
 * "alta simple"), pero el modelo decodifica ambas formas igual — la lista
 * puede traer automatizaciones tipo webhook creadas desde el panel web.
 */
@Serializable
data class AutomationTrigger(
    val kind: String = "schedule",
    val rrule: String? = null,
    @SerialName("has_secret") val hasSecret: Boolean = false,
    @SerialName("hook_url") val hookUrl: String? = null,
)

/** `Automation.accion` — única variante pinned hoy
 * (`AgentInstructionAccion`, `edecan_schemas.automations`). */
@Serializable
data class AutomationAccion(
    val kind: String = "agent_instruction",
    val instruccion: String = "",
    val agente: String? = null,
)

/**
 * Fila pública de `automations` (`GET/POST/PATCH /v1/automations*`).
 */
@Serializable
data class Automation(
    val id: String,
    val nombre: String = "",
    val descripcion: String = "",
    val trigger: AutomationTrigger = AutomationTrigger(),
    val accion: AutomationAccion = AutomationAccion(),
    val enabled: Boolean = true,
    @SerialName("next_run_at") val nextRunAt: String? = null,
    @SerialName("last_run_at") val lastRunAt: String? = null,
    @SerialName("created_at") val createdAt: String = "",
    @SerialName("updated_at") val updatedAt: String = "",
    /** Solo viene poblado en la respuesta puntual del `POST`/`PATCH` que
     * ACABA de generarlo (ver docstring de `_public_automation`) — este
     * cliente no crea automatizaciones tipo webhook (ver
     * `AutomatizacionesScreen`), así que en la práctica siempre llega
     * `null` aquí, pero se modela igual por completitud del contrato. */
    @SerialName("hook_secret") val hookSecret: String? = null,
)

/** Una fila de `GET /v1/automations/{id}/runs`. `detalle` queda como
 * [JsonElement] crudo — su forma depende de qué corrió esa automatización,
 * no está pinned (`automation_runs.detalle`, `ROADMAP_V2.md` §7.4). */
@Serializable
data class AutomationRun(
    val id: String,
    val status: String = "running",
    val detalle: JsonElement? = null,
    @SerialName("started_at") val startedAt: String? = null,
    @SerialName("finished_at") val finishedAt: String? = null,
)

File: apps/mobile/ios/EdecanKit/Sources/EdecanKit/AutomationsModels.swift
import Foundation

// MARK: - `/v1/automations` (`ARCHITECTURE.md` §11/`ROADMAP_V2.md`
// §7.4/§7.6/§7.10, `apps/api/edecan_api/routers/automations.py`) — reglas de
// agenda o webhook que corren una instrucción del agente sin supervisión.
// Este cliente móvil solo CREA disparadores `kind="schedule"` (agenda por
// `rrule` RFC 5545, ver `APIClient.createAutomation`) — dar de alta un
// webhook desde el teléfono queda fuera de alcance de esta app (sigue
// viviendo en el panel web), aunque sí decodifica automatizaciones
// existentes de tipo webhook para poder listarlas/mostrarlas.

/// `AutomationOut.trigger` — unión discriminada por `"kind"`, mismo criterio
/// que ``ChatEvent``: solo `Decodable` (este cliente nunca reenvía un
/// `AutomationOut.trigger` tal cual; para crear una automatización arma su
/// propio cuerpo `{"kind": "schedule", "rrule": ...}` a mano en
/// `APIClient.createAutomation`). `hasSecret`/`hookUrl` de `.webhook` son la
/// forma YA REDACTADA que devuelve `automations.py::_public_automation` — el
/// secreto real (si se acaba de generar) solo viaja en
/// `AutomationOut.hookSecret`, nunca aquí.
public enum AutomationTrigger: Decodable, Sendable, Equatable {
    /// `descripcion` es la `rrule` en cristiano ("Todos los días a las 4:00
    /// UTC"), armada por el SERVIDOR
    /// (`automations.py::_public_automation` → `edecan_automations.describe`)
    /// y nunca acá: la frase lleva HORA, y sólo el backend sabe en qué zona
    /// corre de verdad la regla. Por eso viene rotulada UTC y NO se convierte
    /// a la del teléfono — el instante concreto en hora local ya lo da
    /// ``AutomationOut/nextRunAt`` aparte, que sí es absoluto.
    ///
    /// `nil` cuando el servidor no pudo traducirla con certeza (`UNTIL`, "el
    /// último viernes del mes", …) o cuando es anterior a este campo: en ese
    /// caso se muestra la `rrule` cruda, nunca una hora aproximada — ver
    /// ``resumen``.
    case schedule(rrule: String, descripcion: String?)
    case webhook(hasSecret: Bool, hookUrl: String)

    enum CodingKeys: String, CodingKey {
        case kind, rrule, descripcion
        case hasSecret = "has_secret"
        case hookUrl = "hook_url"
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let kind = try container.decode(String.self, forKey: .kind)
        switch kind {
        case "schedule":
            self = .schedule(
                rrule: try container.decode(String.self, forKey: .rrule),
                descripcion: try container.decodeIfPresent(String.self, forKey: .descripcion)
            )
        case "webhook":
            self = .webhook(
                hasSecret: try container.decodeIfPresent(Bool.self, forKey: .hasSecret) ?? false,
                hookUrl: try container.decodeIfPresent(String.self, forKey: .hookUrl) ?? ""
            )
        default:
            throw DecodingError.dataCorruptedError(
                forKey: .kind, in: container,
                debugDescription: "Trigger de automatización desconocido: \(kind)"
            )
        }
    }

    /// Fallback cuando el servidor no supo traducir la `rrule`: honesto y
    /// corto para una lista (el detalle además enseña la regla cruda).
    /// Espejo de `AGENDA_SIN_TRADUCIR` en
    /// `apps/web/src/lib/api-automatizaciones.ts`.
    public static let agendaSinTraducir = "Agenda personalizada"

    /// Una línea que dice cuándo corre esta automatización. NO arma la frase:
    /// para las de agenda usa la que ya vino del servidor (ver
    /// ``schedule(rrule:descripcion:)``), de modo que el teléfono y la web
    /// digan exactamente lo mismo.
    public var resumen: String {
        switch self {
        case .schedule(_, let descripcion):
            let texto = descripcion?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            return texto.isEmpty ? Self.agendaSinTraducir : texto
        case .webhook:
            return "Se dispara por webhook entrante"
        }
    }
}

/// `AutomationOut.accion` — única variante hoy (`kind == "agent_instruction"`,
/// `edecan_automations.engine.validate_accion`), así que a diferencia de
/// ``AutomationTrigger`` no hace falta una unión discriminada.
public struct AutomationAccion: Decodable, Sendable, Equatable {
    public let kind: String
    public let instruccion: String
    public let agente: String?
}

/// Fila pública de `automations` (`automations.py::_public_automation` — sin
/// un `BaseModel` de Pydantic pinned, el router arma un `dict` a mano, ver su
/// docstring; forma verificada campo por campo contra esa función).
/// `hookSecret` SOLO viene poblado en la respuesta puntual del `POST`/`PATCH`
/// que lo generó — nunca en `GET`/`list`.
///
/// `enabled` es el ÚNICO campo `var` (el resto son `let`, igual criterio que
/// el resto de este archivo): `AutomatizacionesViewModel.alternar` lo muta
/// in-place para el update optimista del toggle, revirtiéndolo si el
/// servidor rechaza el cambio — ver su docstring.
public struct AutomationOut: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let nombre: String
    public let descripcion: String
    public let trigger: AutomationTrigger
    public let accion: AutomationAccion
    public var enabled: Bool
    public let nextRunAt: Date?
    public let lastRunAt: Date?
    public let createdAt: Date
    public let updatedAt: Date
    public let hookSecret: String?

    enum CodingKeys: String, CodingKey {
        case id, nombre, descripcion, trigger, accion, enabled
        case nextRunAt = "next_run_at"
        case lastRunAt = "last_run_at"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case hookSecret = "hook_secret"
    }
}

/// Una corrida de `automation_runs`
/// (`automations.py::list_automation_runs`). `status` ∈
/// `running|done|error|waiting_confirmation` — `String` crudo, mismo
/// criterio que `MissionOut.status`. `detalle` es forma libre
/// (`jsonb`, escrita por `automation_scan.py`/`hooks.py`): puede traer
/// `"resultado"`/`"error"`/`"pendiente"` según cómo terminó la corrida.
public struct AutomationRunOut: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    public let status: String
    public let detalle: [String: JSONValue]
    public let startedAt: Date?
    public let finishedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, status, detalle
        case startedAt = "started_at"
        case finishedAt = "finished_at"
    }
}
